vms/platformvm/txs/mempool/mempool.go

```diff
@@ -4,42 +4,353 @@
 }
 
-func New(namespace string, registerer prometheus.Registerer) (*Mempool, error) {
-	metrics, err := txmempool.NewMetrics(namespace, registerer)
+type meteredTx struct {
+	*txs.Tx
+	// gasPrice is the amount of AVAX burned per unit of gas used by this tx
+	gasPrice float64
+	gasUsed  gas.Gas
+}
+
+type mempool struct {
+	weights     gas.Dimensions
+	avaxAssetID ids.ID
+
+	lock               sync.RWMutex
+	cond               *lock.Cond
+	tree               *btree.BTreeG[meteredTx]
+	txs                map[ids.ID]meteredTx
+	consumedUTXOs      *setmap.SetMap[ids.ID, ids.ID]
+	droppedTxIDs       *lru.Cache[ids.ID, error]
+	gasAvailable       gas.Gas
+	numTxsMetric       prometheus.Gauge
+	gasAvailableMetric prometheus.Gauge
+}
+
+func New(
+	namespace string,
+	weights gas.Dimensions,
+	gasCapacity gas.Gas,
+	avaxAssetID ids.ID,
+	registerer prometheus.Registerer,
+) (Mempool, error) {
+	numTxsMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: namespace,
+		Name:      "count",
+		Help:      "number of transactions in the mempool",
+	})
+
+	gasAvailableMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: namespace,
+		Name:      "gas_available",
+		Help:      "amount of gas available",
+	})
+
+	if err := errors.Join(
+		registerer.Register(numTxsMetric),
+		registerer.Register(gasAvailableMetric),
+	); err != nil {
+		return nil, fmt.Errorf("failed to register metrics: %w", err)
+	}
+
+	m := &mempool{
+		weights:     weights,
+		avaxAssetID: avaxAssetID,
+		tree: btree.NewG[meteredTx](2, func(a, b meteredTx) bool {
+			if a.gasPrice != b.gasPrice {
+				return a.gasPrice < b.gasPrice
+			}
+
+			// Break ties with txID
+			return a.TxID.Compare(b.TxID) < 0
+		}),
+		txs:                make(map[ids.ID]meteredTx),
+		consumedUTXOs:      setmap.New[ids.ID, ids.ID](),
+		droppedTxIDs:       lru.NewCache[ids.ID, error](64),
+		gasAvailable:       gasCapacity,
+		numTxsMetric:       numTxsMetric,
+		gasAvailableMetric: gasAvailableMetric,
+	}
+
+	m.cond = lock.NewCond(&m.lock)
+	return m, nil
+}
+
+func (m *mempool) Add(tx *txs.Tx) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.consumedUTXOs.HasOverlap(tx.InputIDs()) {
+		return txmempool.ErrConflictsWithOtherTx
+	}
+
+	meteredTx, err := m.meter(tx)
+	if err != nil {
+		return fmt.Errorf("failed to meter tx: %w", err)
+	}
+
+	// Try to evict lower gas priced txs if we do not have enough remaining gas
+	// capacity
+	if err := m.allocateSpace(meteredTx); err != nil {
+		return err
+	}
+
+	m.tree.ReplaceOrInsert(meteredTx)
+	m.txs[meteredTx.TxID] = meteredTx
+	m.consumedUTXOs.Put(meteredTx.TxID, meteredTx.InputIDs())
+	m.droppedTxIDs.Evict(meteredTx.TxID)
+	m.gasAvailable -= meteredTx.gasUsed
+
+	m.updateMetrics()
+	m.cond.Broadcast()
+
+	return nil
+}
+
+// Try to evict transactions until there is enough capacity for the tx.
+// Returns if any txs were evicted.
+func (m *mempool) allocateSpace(txToAdd meteredTx) error {
```

Comment:
nit: I feel like this function should just return a bool

Reply:
I was dumb and in the previous review incorrectly thought this had to be an error for some reason. Returning a bool now.